### PR TITLE
[diffusion] feat: support efficient sequence shard

### PR DESCRIPTION
**Pull request by nono-Sang**
## Motivation

For video models, when SGLang enables SP (Sequence Parallelism), it uses frame-level partitioning. The number of frames needs to be adjusted to meet the divisibility condition, and the number of padding tokens is `pad_frames * latent_H * latent_W`. However, we can directly partition along the `frames * latent_H * latent_W` dimension, significantly reducing the number of padding tokens and improving computational efficiency.

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review Process

1. Ping Merge Oncalls to start the PR flow. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - `/tag-run-ci-label`, `/rerun-failed-ci`, `/tag-and-rerun-ci`
4. After green CI and required approvals, ask Merge Oncalls to merge.

### CODE REVIEW

File: python/sglang/multimodal_gen/configs/pipeline_configs/base.py [hunk 1]
```diff
@@ -216,6 +216,9 @@ def postprocess_image(self, image):
     # Compilation
     # enable_torch_compile: bool = False
 
+    # "frame" shard by `frame` dimension, "sequence" shard by `frame * H * W` dimension
+    sp_shard_mode: str = "frame"
```
Comment (mickqian) [inline python/sglang/multimodal_gen/configs/pipeline_configs/base.py]: why is this field is defined in `PipelineConfig`, while accessed via `Req`?

File: python/sglang/multimodal_gen/configs/pipeline_configs/base.py [hunk 2]
```diff
@@ -340,6 +343,8 @@ def preprocess_vae_image(self, batch, vae_image_processor):
 
     def shard_latents_for_sp(self, batch, latents):
         # general logic for video models
+        if getattr(batch, "sp_shard_mode", "frame") == "sequence":
```
Comment (mickqian) [inline python/sglang/multimodal_gen/configs/pipeline_configs/base.py]: we don't need `getattr` here, since the runtime type is deterministic

File: python/sglang/multimodal_gen/runtime/models/dits/wanvideo.py [hunk 3]
```diff
@@ -805,25 +820,79 @@ def forward(
...
+                )
+                hidden_states = torch.cat([hidden_states, pad], dim=1)
+            sp_rank = get_sp_group().rank_in_group
+            local_seq_len = hidden_states.shape[1] // self.sp_size
+            hidden_states = hidden_states.view(
+                batch_size, self.sp_size, local_seq_len, hidden_states.shape[2]
+            )
+            hidden_states = hidden_states[:, sp_rank, :, :]
+
+            def _compute_rope_for_sequence_shard(
```
Comment (mickqian) [inline python/sglang/multimodal_gen/runtime/models/dits/wanvideo.py]: please avoid local function def

Comment (mickqian) [reply python/sglang/multimodal_gen/runtime/models/dits/wanvideo.py]: also, if I'm understanding this correctly, it could be cached with `@lru_cache(maxsize=1)` across timesteps. Moving it to `PipelineConfig.get_freqs_cis` is also a decent approach

Comment (nono-Sang) [reply python/sglang/multimodal_gen/runtime/models/dits/wanvideo.py]: @mickqian done

**Review by mickqian — APPROVED** [review]

File: python/sglang/multimodal_gen/configs/sample/sampling_params.py [hunk 4]
```diff
@@ -154,6 +154,7 @@ class SamplingParams:
     suppress_logs: bool = False
 
     return_file_paths_only: bool = True
+    enable_sequence_shard: bool = False
```
Comment (mickqian) [inline python/sglang/multimodal_gen/configs/sample/sampling_params.py:157]: should we make it default enabled?
